jwst/wfss_contam/wfss_contam.py

```diff
@@ -26,71 +25,159 @@ def contam_corr(input_model, waverange, photom):
+
+    # Create simulated grism image for each order and sum them up
+    for order in spec_orders:
+
+        log.info(f"Creating full simulated grism image for order {order}")
+        obs.disperse_all(order, wmin[order], wmax[order], sens_waves[order],
+                         sens_response[order])
+
+        # Accumulate result for this order into the combined image
+        if simul_all is None:
+            simul_all = obs.simulated_image
+        else:
+            simul_all += obs.simulated_image
+
+    # Save the full-frame simulated grism image
     simul_model = datamodels.ImageModel(data=simul_all)
-    root, _ = splitext(input_model.meta.filename)
-    simul_model.meta.filename = replace_suffix(root, 'simul') + '.fits'
-    log.info(f"Saving full simulated image as {simul_model.meta.filename}")
-    simul_model.save(simul_model.meta.filename)
+    simul_model.update(input_model, only="PRIMARY")
 
-    # Loop over all slits/sources
-    log.debug("Creating contam image for each individual source")
+    # Loop over all slits/sources to subtract contaminating spectra
+    log.info("Creating contamination image for each individual source")
+    contam_model = datamodels.MultiSlitModel()
+    contam_model.update(input_model)
+    slits = []
     for slit in output_model.slits:
 
         # Create simulated spectrum for this source only
         sid = slit.source_id
         order = slit.meta.wcsinfo.spectral_order
-        if order != 1:
-            continue
-        obs.disperse_chunk(sid)
+        chunk = np.where(obs.IDs == sid)[0][0]  # find chunk for this source
+
+        obs.simulated_image = np.zeros(obs.dims)
+        obs.disperse_chunk(chunk, order, wmin[order], wmax[order],
+                           sens_waves[order], sens_response[order])
         this_source = obs.simulated_image
 
-        # Contamination estimate is full simulated image
-        # minus this source
+        # Contamination estimate is full simulated image minus this source
         contam = simul_all - this_source
 
-        # Subtract the cutout of the contam image from the slit image
+        # Create a cutout of the contam image that matches the extent
+        # of the source slit
         x1 = slit.xstart - 1
         x2 = x1 + slit.xsize
         y1 = slit.ystart - 1
         y2 = y1 + slit.ysize
-        slit.data -= contam[y1:y2, x1:x2]
+        cutout = contam[y1:y2, x1:x2]
+        new_slit = datamodels.SlitModel(data=cutout)
+        copy_slit_info(slit, new_slit)
+        slits.append(new_slit)
+
+        # Subtract the cutout from the source slit
+        slit.data -= cutout
+
+    # Save the contamination estimates for all slits
+    contam_model.slits.extend(slits)
 
     # Set the step status to COMPLETE
     output_model.meta.cal_step.wfss_contam = 'COMPLETE'
 
-    return output_model
+    return output_model, simul_model, contam_model
+
+
+def copy_slit_info(input_slit, output_slit):
```

Comment:
should this be a method of the model (SlitModel)?

Reply:
That would be convenient, because it's needed in at least a couple different steps in the pipeline (e.g. transferring slit meta data from 2D cutouts to 1D spectra in extract_1d). I tried using `output.update(input, only="SCI")`, but that had no effect, for some reason. Different observing modes have different subsets of the slit meta data populated, but that could be handled in a generic method by only copying those items that exist in the input.

Reply:
This is because of #527.  `update()` only works on things in `meta`.

Reply:
Ah, of course. And of the slit "meta" data is not within the actual `meta` tree. Another good reason to move it all into `meta` (some day).

Reply:
We could do that next week (if its prioritized).  😉 

We could also move the gWCS attribute `wcs` out of `meta`, as it doesn't belong there either.

Reply:
The main concern with finally making that move is whether and how much existing code base (outside of the pipeline) might get broken.

Reply:
We could leave behind a pointer (i.e. keep it working) and issue a `DeprecationWarning`.  And leave it like that for a while.  But then at least internally `update` would work as we expect, and we could remove a bunch of duplicate code that tries to work around this.